### PR TITLE
trace {open, exec}: Revert map sizes to their previous value.

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

This PR revert map sizes to their previous values of 10240 entries in case the memory hungry options are not used.
Indeed, having only 1024 entries caused vertical scaling problem when run on big machine:

```
francis-ig-vertical@francis-ig-verticalvm:~$ sudo ./ig-bad trace exec -o json -c test-container > /tmp/ig-bad &[1] 2060496
francis-ig-vertical@francis-ig-verticalvm:~$ docker run --user 1000 -ti --rm --name test-container --entrypoint '/bin/sh' -w /tmp ghcr.io/colinianking/stress-ng:master -c 'stress-ng --exec $(nproc) --exec-fork-method clone --exec-method execve --exec-no-pthread --timeout 1s --metrics-brief'
stress-ng: info:  [1] setting to a 1 second run per stressor
stress-ng: info:  [1] dispatching hogs: 96 exec
stress-ng: info:  [1] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [1]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [1] exec              42213      2.90     34.04    214.31     14567.68         169.98
stress-ng: info:  [1] successful run completed in 2.99s
francis-ig-vertical@francis-ig-verticalvm:~$ wc -l /tmp/ig-bad 
11604 /tmp/ig-bad
francis-ig-vertical@francis-ig-verticalvm:~$ fg
sudo ./ig-bad trace exec -o json -c test-container > /tmp/ig-bad
francis-ig-vertical@francis-ig-verticalvm:~$ sudo ./ig trace exec -o json -c test-container > /tmp/ig &[1] 2103140
francis-ig-vertical@francis-ig-verticalvm:~$ docker run --user 1000 -ti --rm --name test-container --entrypoint '/bin/sh' -w /tmp ghcr.io/colinianking/stress-ng:master -c 'stress-ng --exec $(nproc) --exec-fork-method clone --exec-method execve --exec-no-pthread --timeout 1s --metrics-brief'
stress-ng: info:  [1] setting to a 1 second run per stressor
stress-ng: info:  [1] dispatching hogs: 96 exec
stress-ng: info:  [1] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [1]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [1] exec              52117      3.49     44.41    251.56     14944.49         176.09
stress-ng: info:  [1] successful run completed in 3.59s
francis-ig-vertical@francis-ig-verticalvm:~$ wc -l /tmp/ig
52120 /tmp/ig
francis-ig-vertical@francis-ig-verticalvm:~$ nproc
96
```

When the memory hungry options are used, the size remains 1024 entries:

```bash
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ sudo ./ig trace exec &                                francis/increase-map-sizes *+%
[1] 115471
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ RUNTIME.CONTAINERNAME          PID              PPID             COMM             RET ARGS
sudo bpftool map list | grep -C 5 execs
        key 8B  value 4B  max_entries 1024  memlock 16384B
        pids ig(115477)
266: array  name .rodata  flags 0x80
        key 4B  value 7768B  max_entries 1  memlock 8192B
        btf_id 388  frozen
267: hash  name execs  flags 0x0
        key 4B  value 7752B  max_entries 10240  memlock 79462400B
        btf_id 389
        pids ig(115477)
268: perf_event_array  name events  flags 0x0
        key 4B  value 4B  max_entries 8  memlock 4096B
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ sudo ./ig trace exec --cwd &                          francis/increase-map-sizes *+%
[1] 115561
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ RUNTIME.CONTAINERNAME          PID              PPID             COMM             RET ARGS                                     CWD
sudo bpftool map list | grep -C 5 execs
        key 8B  value 4B  max_entries 1024  memlock 16384B
        pids ig(115567)
286: array  name .rodata  flags 0x80
        key 4B  value 11864B  max_entries 1  memlock 12288B
        btf_id 418  frozen
287: hash  name execs  flags 0x0
        key 4B  value 11848B  max_entries 1024  memlock 12140544B
        btf_id 419
        pids ig(115567)
288: percpu_array  name bufs  flags 0x0
        key 4B  value 32768B  max_entries 1  memlock 266240B
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ sudo ./ig trace open --full-path &                    francis/increase-map-sizes *+%
[1] 115650
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ RUNTIME.CONTAINERNAME          PID              COMM             FD  ERR PATH                             FULLPATH
sudo bpftool map list | grep -C 5 start
        key 8B  value 4B  max_entries 1024  memlock 16384B
        pids ig(115656)
307: array  name .rodata  flags 0x80
        key 4B  value 4448B  max_entries 1  memlock 8192B
        btf_id 449  frozen
308: hash  name start  flags 0x0
        key 4B  value 4416B  max_entries 1024  memlock 4530176B
        btf_id 450
        pids ig(115656)
309: perf_event_array  name events  flags 0x0
        key 4B  value 4B  max_entries 8  memlock 4096B
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ sudo ./ig trace open &                                francis/increase-map-sizes *+%
[2] 115826
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ RUNTIME.CONTAINERNAME          PID              COMM             FD  ERR PATH
francis@pwmachine:~/Codes/kinvolk/inspektor-gadget$ sudo bpftool map list | grep -C 5 start               francis/increase-map-sizes *+%
        key 8B  value 4B  max_entries 1024  memlock 16384B
        pids ig(115776)
332: array  name .rodata  flags 0x80
        key 4B  value 4448B  max_entries 1  memlock 8192B
        btf_id 491  frozen
333: hash  name start  flags 0x0
        key 4B  value 4416B  max_entries 10240  memlock 45301760B
        btf_id 492
        pids ig(115776)
334: perf_event_array  name events  flags 0x0
        key 4B  value 4B  max_entries 8  memlock 4096B
```

Best regards.

### CODE REVIEW

File: pkg/gadgets/trace/open/tracer/tracer.go [hunk 1]
```diff
@@ -101,6 +101,10 @@ func (t *Tracer) install() error {
 	consts := make(map[string]interface{})
 	consts["get_full_path"] = t.config.FullPath
 
+	if t.config.FullPath {
+		spec.Maps["start"].MaxEntries = 1024
+	}
+
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/open/tracer/tracer.go:107]: I don't think it'll work as the structure always has `full_fname` defined:

https://github.com/inspektor-gadget/inspektor-gadget/blob/5438f5e9e0c9f2cbe781a1eb288f7765a7634806/pkg/gadgets/trace/open/tracer/bpf/opensnoop.h#L10C1-L24C3

I think an `#ifdef` logic is needed here. An  ["Array of Length Zero"](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html) could be a more elegant solution, but I don't know if clang and the verifier will be happy with it.

**Comment by mauriciovasquezbernal** [comment]
After checking it with more care, I think we can handle it in a different way in trace open.

TL:DR: We're saving a lot of data we don't need in the start map

The trace map is used to keep context between the trace_enter and trace_exit functions:

https://github.com/inspektor-gadget/inspektor-gadget/blob/d7c9ae4aa7de0bf2c366933acc8e10157cc1cf5b/pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c#L25-L30

We save the full event structure there

https://github.com/inspektor-gadget/inspektor-gadget/blob/d7c9ae4aa7de0bf2c366933acc8e10157cc1cf5b/pkg/gadgets/trace/open/tracer/bpf/opensnoop.h#L10-L24

But only few member are written in the trace_enter function

https://github.com/inspektor-gadget/inspektor-gadget/blob/d7c9ae4aa7de0bf2c366933acc8e10157cc1cf5b/pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c#L150-L152

Everything thing else we keep in that structure is a waste of memory, since we don't need to store them in the map as they're filled in the trace_exit function.

I did an experiment to define a smaller structure with only the needed data (available [here](https://github.com/inspektor-gadget/inspektor-gadget/commit/2854ff82a4567e2def6698abb342fab6aa21f98e)) and the results are good. This is a summary of different situations:

main branch (1024 entries, 4.3MB)
```
82: hash  name start  flags 0x0
	key 4B  value 4416B  max_entries 1024  memlock 4530176B
	btf_id 320
	pids ig(45749)
```

This PR with --full-path (1024 entries, 4.3MB)

```
308: hash  name start  flags 0x0
        key 4B  value 4416B  max_entries 1024  memlock 4530176B
        btf_id 450
        pids ig(115656)
```

This PR without --full-path (10240 entries, 3.2MB)

```
126: hash  name start  flags 0x0
	key 4B  value 320B  max_entries 10240  memlock 3358720B
	btf_id 384
	pids ig(47161)
```

Mauricio's idea independent of --full-path value. (10240 entries, 2.6MB)

```
59: hash  name start  flags 0x0
	key 4B  value 264B  max_entries 10240  memlock 2785280B
	btf_id 287
	pids ig(45299)
```

There is an an additional map needed by this solution, but that's only 136KB on a 16 cores machine.

```
149: percpu_array  name empty_event  flags 0x0
	key 4B  value 4416B  max_entries 1  memlock 143360B
	btf_id 418
	pids ig(48543)
```

In trace exec we can't do the same as `cwd` is saved on the "enter" phase. Perhaps that can be done on the exit phase too?

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM! 

Would you mind opening an issue to check if we can optimize trace exec on the same way? 

Thanks!
